File: nim/src/jsonpointer.nim
## JSON Pointer (RFC 6901) — parse + resolve. Mirror of python/jsoncanon/
## jsonpointer.py. Shared by --pointer and JSON Patch.

import std/strutils
import ./jsoncanon

proc ptrUnescape*(tok: string): string =
  tok.replace("~1", "/").replace("~0", "~")

proc parsePointer*(pt: string): seq[string] =
  if pt.len == 0: return @[]
  if pt[0] != '/':
    raise newException(JsonError, "invalid JSON Pointer '" & pt & "' (must be empty or start with '/')")
  for t in pt[1 .. ^1].split('/'): result.add ptrUnescape(t)

proc arrayIndex*(tok: string, length: int, allowDash = false): int =
  if allowDash and tok == "-": return length
  if tok != "0" and (tok.len == 0 or tok[0] == '0' or (block:
       var ok = true
       for c in tok: (if c notin {'0'..'9'}: ok = false)
       not ok)):
    raise newException(JsonError, "JSON Pointer: invalid array index '" & tok & "'")
  parseInt(tok)

proc resolveTokens*(doc: JNode, toks: seq[string]): JNode =
  result = doc
  for tok in toks:
    case result.kind
    of jkObj:
      let i = result.keys.find(tok)
      if i < 0: raise newException(JsonError, "JSON Pointer: no member '" & tok & "'")
      result = result.vals[i]
    of jkArr:
      let idx = arrayIndex(tok, result.arr.len)
      if idx >= result.arr.len: raise newException(JsonError, "JSON Pointer: index " & tok & " out of range")
      result = result.arr[idx]
    else:
      raise newException(JsonError, "JSON Pointer: cannot descend into a scalar at '" & tok & "'")

proc getPointer*(doc: JNode, pt: string): JNode =
  resolveTokens(doc, parsePointer(pt))

File: python/jsoncanon/jsonpointer.py
"""JSON Pointer (RFC 6901) — parse + resolve. Shared by --pointer and JSON Patch.

`--pointer EXPR` extracts the referenced sub-value, which is then canonicalized.
Array indices follow RFC 6901: "0" or a no-leading-zero decimal (and "-" only in
JSON Patch `add`, handled there).
"""

from __future__ import annotations

from .parser import JValue, JSONError


def unescape(tok: str) -> str:
    return tok.replace("~1", "/").replace("~0", "~")


def parse_pointer(ptr: str) -> list[str]:
    if ptr == "":
        return []
    if not ptr.startswith("/"):
        raise JSONError(f"invalid JSON Pointer {ptr!r} (must be empty or start with '/')")
    return [unescape(t) for t in ptr[1:].split("/")]


def _array_index(tok: str, length: int, *, allow_dash: bool = False) -> int:
    if allow_dash and tok == "-":
        return length
    if tok != "0" and (tok == "" or tok[0] == "0" or not tok.isdigit()):
        raise JSONError(f"JSON Pointer: invalid array index {tok!r}")
    return int(tok)


def resolve_tokens(doc: JValue, toks: list[str]) -> JValue:
    node = doc
    for tok in toks:
        if isinstance(node, dict):
            if tok not in node:
                raise JSONError(f"JSON Pointer: no member {tok!r}")
            node = node[tok]
        elif isinstance(node, list):
            i = _array_index(tok, len(node))
            if i >= len(node):
                raise JSONError(f"JSON Pointer: index {tok} out of range")
            node = node[i]
        else:
            raise JSONError(f"JSON Pointer: cannot descend into a scalar at {tok!r}")
    return node


def get_pointer(doc: JValue, ptr: str) -> JValue:
    return resolve_tokens(doc, parse_pointer(ptr))
